### ISSUE TITLE
Fail to rewire named exported constant functions

### ISSUE THREAD

**Issue by InakiAbete-Reflektion** [asapach/babel-plugin-rewire-exports#6]
Modules with named exported constant functions like this one are not being "rewired":
```javascript
export const say = str => console.log(str);
```
BTW, it works for named exported functions:
```javascript
export function say() { console.log(str) };
```

**Comment by asapach** [comment]
Yes, it's one of the design limitations: https://github.com/asapach/babel-plugin-rewire-exports#transformations

You can also try this:
```js
const say = str => console.log(str);
export {say};
```

In order to rewire constant exports you'd need to transform it to something like that:
```js
const say = str => console.log(str);
var _say = say;
export { _say as say };
```
So the issue is: should we replace all references of `say` to `_say`?
* If yes, than it's no longer a constant - which will break the code that expects otherwise.
* If not, then the code *inside* the module will not get the rewired value.

Makes sense? How would you expect it to work?

**Comment by asapach** [comment]
Although I think in this case we don't replace any references [(example)](https://github.com/asapach/babel-plugin-rewire-exports/blob/master/test/fixtures/transform-named-export-constant/expected.js):
```js
const foo = 'bar';
export {foo};
```
So essentially the *exported* value/binding is rewired, but the *constant* itself is not, so the code that depends on `foo` inside the module will never see the rewired value.

I guess it's possible to treat `export const foo = 'bar'` in a similar manner.

Thoughts?

**Comment by asapach** [comment]
@InakiAbete-Reflektion ping.

**Comment by InakiAbete-Reflektion** [comment]
I see. So, what you are saying is that do this wouldn't be convenient because `foo` won't be a constant anymore, right?

```javascript
export const foo = str => console.log(str);
```

```javascript
var _foo = str => console.log(str);
var foo = _foo;
export { foo };
export function rewire$foo($stub) {
  foo = $stub;
}
export function restore() {
  foo = _foo;
}
```

**Comment by asapach** [comment]
Yeah, pretty much. But I'm not concerned about *convenience* -- rather about *correctness*: by enabling this you change your code in a way that modifies its behavior. And I've tried to avoid that whenever possible.

**Comment by asapach** [comment]
I think this should preserve correctness:
```js
const foo = str => console.log(str); // still const
var _foo = foo; // just a local var
export { _foo as foo };
export function rewire$foo($stub) {
  _foo = $stub;
}
export function restore() {
  _foo = foo;
}
```
But as you can see, you can only rewire `_foo` which is exported, but not `foo` which might be used inside the module.

**Comment by InakiAbete-Reflektion** [comment]
I don’t see how it’s different from named constants like:

```
export const foo = true;
export function bar() { console.log(foo); }
```
that can be used inside the module.

**Comment by asapach** [comment]
You mean like that?
```js
const foo = true;
export {foo};
export function bar() { console.log(foo); }
```
In this case `foo` will not really be rewired. Only the exported value will.

Would you expect these two to work the same?
```js
const foo = true;
export {foo};
```
vs
```js
export const foo = true;
```

**Comment by InakiAbete-Reflektion** [comment]
I’m still trying to understand this, sorry 😊

Wouldn’t this work?
`export const foo = true`
Even if it is used inside the module?

If that works, I don’t get why `export const foo = s => console.log(s);`
doesn’t

**Comment by asapach** [comment]
Nope, it wouldn't work because I explicitly ignore this scenario: https://github.com/asapach/babel-plugin-rewire-exports/blob/master/src/index.js#L141

**Comment by InakiAbete-Reflektion** [comment]
Now I get it. Thanks.
So there's no way to solve it without changing the definition, right? That's sad.

**Comment by asapach** [comment]
Correct. You need to separate the constant from the export one way or another.
Alternatively I could add a plugin option that would allow you to rewire constants, but it will not be enabled by default, because it might break your code.

**Comment by InakiAbete-Reflektion** [comment]
That could be a good option to have

**Comment by itaylor** [comment]
I'm looking to migrate a large codebase that currently uses `babel-plugin-rewire` to this tool.  Given the prevalence of `export const` usage, I think it might make sense to add an option that basically transforms `export const foo` to `export let foo` which could then allow for `foo` to be rewired later.  Converting a `const` to a `let` is still far less impactful than what `babel-plugin-rewire` is doing.

**Comment by asapach** [comment]
Fixed in 1dcf35ec39f9c883d6ea53b67afceb249112ac75: added [`unsafeConst` option](https://github.com/asapach/babel-plugin-rewire-exports#unsafeconst)
Released as `v0.4.0` and `v1.0.0-beta` (for Babel v7)
So when the option is turned on, this:
```js
export const foo = 'bar'
```
Becomes this:
```js
export let foo = 'bar'
```

It however does not affect any of these exports:
```js
const foo = "bar"
export {foo}
export default foo
```
In both of these cases you can rewire the exports, but `foo` remains const and you can't change it.
